File: Leetcode/Trapping Rain Water.cc
class Solution {
public:
    int trap(int A[], int n) {
        int MaxLeft[n], MaxRight[n];
        memset(MaxLeft, 0, sizeof(MaxLeft));
        memset(MaxRight, 0, sizeof(MaxRight));
        
        for (int i = 0; i < n; i++) {
            if (i) MaxLeft[i] = max(A[i], MaxLeft[i-1]);
            else MaxLeft[i] = A[i];
        }
        
        for (int i = n - 1; i >= 0; i--) {
            if (i + 1 < n) MaxRight[i] = max(A[i], MaxRight[i+1]);
            else MaxRight[i] = A[i];
        }
        
        int ans = 0;
        for (int i = 0; i < n; i++) {
            int ml = i - 1 >= 0? MaxLeft[i-1] : 0;
            int mr = i + 1 < n? MaxRight[i+1] : 0;
            int h = min(ml, mr) - A[i];
            ans += h > 0? h : 0;
        }
        return ans;
    }
};

File: Leetcode/Algorithm/java/Trapping Rain Water.java
public class Solution {
    public int trap(int[] A) {
        int n = A.length;
        int max_left = 0, max_right = 0;
        int p = 0, q = n - 1;
        int ans = 0;
        while (p < q) {
            if (A[p] < A[q]) {
                max_left = Math.max(max_left, A[p]);
                ans += max_left - A[p];
                p++;
            } else {
                max_right = Math.max(max_right, A[q]);
                ans += max_right - A[q];
                q--;
            }
        }
        return ans;
    }
}